logr.go

```diff
@@ -140,24 +140,20 @@ import (
 	"context"
 )
 
-// TODO: consider adding back in format strings if they're really needed
-// TODO: consider other bits of zap/zapcore functionality like ObjectMarshaller (for arbitrary objects)
-// TODO: consider other bits of glog functionality like Flush, OutputStats
-
-// Logger represents the ability to log messages, both errors and not.
-type Logger interface {
-	// Enabled tests whether this Logger is enabled.  For example, commandline
+// LogSink represents the ability to log messages, both errors and not.
+type LogSink interface {
+	// Enabled tests whether this LogSink is enabled.  For example, commandline
 	// flags might be used to set the logging verbosity and disable some info
 	// logs.
-	Enabled() bool
+	Enabled(level int) bool
```

Comment:
At this moment it's only the level, but perhaps we would like to pull some other information into the concrete `Logger` in the future, which would require breaking changes to pass it to the `Logger`.

One alternative would be to pass it in a struct:

```go
type LogContext struct {
    level int
}
```

And then change the interface to:

```go
Enabled(logctx LogContext) bool
Info(logctx LogContext, msg string, keysAndValues ...interface{})
...
```

Reply:
Perhaps this `LogContext` can also contain the number of stack frames added by the log stack. This would allow the creation of middleware-like log sinks that delegate to other sinks, while keeping the call stack count correct.

Reply:
> Perhaps this LogContext can also contain the number of stack frames added by the log stack. This would allow the creation of middleware-like log sinks that delegate to other sinks, while keeping the call stack count correct.

This is what `WithCallDepth()` is for.  I agree that assuming literal `1` in the LogSink is kind of a mess, though.  Passing it to every call to `Info()` feels a bit silly - it's not going to change, like the levels.  It almost feels like `logr.CallFrameCount()` (simple, ugly) or something like:

`logr.New(sink)` calls `sink.Init(numCallFrames)`, (probably a struct) which can choose to store that info if it needs.